### PR TITLE
RRE on AWS

### PR DESCRIPTION
**Pull request by mathcolo**
This PR patches RRE to deploy end-to-end to a fresh AWS EC2 instance, with automatic DNS record configuration via a labs.transitmatters.org subdomain.

Please consult the (WIP!) documentation at https://docs.google.com/document/d/1WjCsdzGlIQjt26CYvinbA2-yIXOAuxmTz8GYgOYS3MI/edit before reviewing this PR, since it will likely answer some architectural questions about how it works.

Live ~beta URL: https://rre-beta.labs.transitmatters.org~ at https://regionalrail.rocks

### CODE REVIEW

File: devops/deploy-on-ec2.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+---
+- name: deploy RRE on EC2
+  hosts: all
+  remote_user: ubuntu
+  become_user: ubuntu
+  tasks:
+    - name: "Add nodejs apt key"
```
Comment (nathan-weinberg) [inline devops/deploy-on-ec2.yml:7]: Can we remove quotes here for consistency with other tasks?

File: devops/deploy-on-ec2.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      become: yes
+      become_user: root
+      apt_key:
+        url: https://deb.nodesource.com/gpgkey/nodesource.gpg.key
+        state: present
+
+    - name: "Add nodejs 16.x ppa for apt repo"
```
Comment (nathan-weinberg) [inline devops/deploy-on-ec2.yml:14]: Can we remove quotes here for consistency with other tasks?

File: devops/deploy-on-ec2.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    - name: build application
+      shell: npm run build
+      args:
+        chdir: /home/ubuntu/regional-rail-explorer
+
+    - name: restart service
+      shell: systemctl restart regional-rail-explorer
+      become: yes
+      become_user: root
```
Comment (nathan-weinberg) [inline devops/deploy-on-ec2.yml:74]: Can we move 
```
      become: yes
      become_user: root
```
directly under the task name where applicable for consistency?

File: devops/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            CI=true
+            ;;
+  esac
+done
+
+$PRODUCTION && HOSTNAME="regionalrail.rocks" || HOSTNAME="rre-beta.labs.transitmatters.org"
+$PRODUCTION && DOMAIN="regionalrail.rocks" || DOMAIN="labs.transitmatters.org"
+$PRODUCTION && CERT_ARN="$TM_RRE_RR_RKS_CERT_ARN" || CERT_ARN="$TM_LABS_WILDCARD_CERT_ARN"
+
+# STACK_NAME="rre-stack-beta-preston"
```
Comment (nathan-weinberg) [inline devops/deploy.sh:31]: Let's remove :)

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]
Thanks for doing this Preston it looks great! Some stylistic requests but otherwise I have no problems with this.

As for CD, this could be changed to mimic [what we're doing for the Data Dashboard](https://github.com/transitmatters/t-performance-dash/blob/main/.github/workflows/deploy.yml) - it might actually be even simpler since a lot of the dependency steps won't be necessary here.

**Review by nathan-weinberg — APPROVED** [review]
LGTM! Can help with followup for CD if one is needed - thanks again Preston!

**Comment by idreyn** [comment]
https://www.youtube.com/watch?v=QPM6FFD2eCk

**Comment by mathcolo** [comment]
@idreyn Not merged yet, sorry lol

**Comment by idreyn** [comment]
(angry air horn noises)

File: devops/cloudformation.json [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      "Properties": {
+        "GroupName": {"Fn::Sub": "RRE-securitygroup-${RREHostname}"},
+        "GroupDescription": "Allow HTTP/HTTPS and SSH inbound and outbound traffic",
+        "SecurityGroupIngress": [
+          {
+            "IpProtocol": "tcp",
+            "FromPort": 80,
+            "ToPort": 80,
+            "CidrIp": "0.0.0.0/0"
+          },
```
Comment (austinjpaul) [inline devops/cloudformation.json:49]: Do we actually need to allow HTTP/80 from the whole internet? Won't all these requests just come from Route 53? Or is Route53 not a full proxy? I'm a bit shaky on how the ssl layer works.

File: devops/deploy-on-ec2.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      git:
+        repo: https://github.com/transitmatters/regional-rail-explorer.git
+        dest: /home/ubuntu/regional-rail-explorer
+        force: yes
+
+    - name: copy systemd service file
+      become: yes
+      become_user: root
+      copy:
+        src: systemd.conf
```
Comment (austinjpaul) [inline devops/deploy-on-ec2.yml:37]: again, somehow it knows we're in the devops folder?

**Review by austinjpaul — APPROVED** [review]
I think I understand about 85% of this, and it's amazing, Well done @mathcolo !!
I left a couple very minor stylistic notes, and a few questions about how things work.

File: devops/cloudformation.json (hunk 5 above)
Comment (austinjpaul) [reply devops/cloudformation.json:49]: Doh, just realized cloudfront is doing this part for us. same question there though

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -8,17 +8,30 @@ on:
...
-            [all]
-            ${{ secrets.LIGHTSAIL_PUBLIC_IP }} ansible_python_interpreter="/usr/bin/python3"
-          options: |
-            --verbose
+          node-version: 16
+      - name: Generate AWS profile
+        run: |
+          mkdir ~/.aws
+          cat >> ~/.aws/credentials << EOF
+          [$AWS_PROFILE]
```
[the remark this answers is not in the document]
Comment (mathcolo) [reply .github/workflows/deploy.yml:25]: Oof, thanks for catching this!

File: devops/deploy.sh [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+aws cloudformation deploy --stack-name $STACK_NAME \
+  --template-file cloudformation.json \
+  --capabilities CAPABILITY_NAMED_IAM \
+  --no-fail-on-empty-changeset \
+  --parameter-overrides \
+  RREHostname=$HOSTNAME \
+  RREDomain=$DOMAIN \
+  RRECertArn=$CERT_ARN
+
+INSTANCE_IP=$(aws cloudformation describe-stacks --stack-name $STACK_NAME --query "Stacks[0].Outputs[?OutputKey=='InstanceIP'].OutputValue" --output text)
```
[the remark this answers is not in the document]
Comment (mathcolo) [reply devops/deploy.sh:45]: set -e is on the script, so if the AWS CLI command fails, this won't get run.

If you mean checking at a different layer that things are running, that could be a nice enhancement!

File: devops/deploy.sh [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+# Run the playbook! :-)
+export ANSIBLE_HOST_KEY_CHECKING=False # If it's a new host, ssh known_hosts not having the key fingerprint will cause an error. Silence it
+ansible-playbook -i $INSTANCE_IP, -u ubuntu --private-key ~/.ssh/transitmatters-rre.pem deploy-on-ec2.yml
```
[the remark this answers is not in the document]
Comment (mathcolo) [reply devops/deploy.sh:48]: 1. nope
2. sure!

File: devops/deploy-on-ec2.yml (hunk 6 above)
Comment (mathcolo) [reply devops/deploy-on-ec2.yml:37]: yeah, it's local to the playbook

File: devops/cloudformation.json (hunk 5 above)
Comment (mathcolo) [reply devops/cloudformation.json:49]: Route 53 is just Amazon's DNS control panel; my doc may have made that unclear.

But yeah, requests come through CloudFront. It appears permitting CloudFront-only traffic is tough, so I figure we can just leave this for now.

File: .github/workflows/deploy.yml [hunk 10]
```diff
@@ -8,17 +8,30 @@ on:
...
+          aws_access_key_id = ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws_secret_access_key = ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          EOF
```
[the remark this answers is not in the document]
Comment (mathcolo) [reply .github/workflows/deploy.yml:28]: @nathan-weinberg and I were considering doing organization secrets. That said, I like the idea of giving this its own IAM, so I went ahead and did that.

File: devops/deploy.sh [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            CI=true
+            ;;
+  esac
+done
+
+$PRODUCTION && HOSTNAME="regionalrail.rocks" || HOSTNAME="rre-beta.labs.transitmatters.org"
+$PRODUCTION && DOMAIN="regionalrail.rocks" || DOMAIN="labs.transitmatters.org"
+$PRODUCTION && CERT_ARN="$TM_RRE_RR_RKS_CERT_ARN" || CERT_ARN="$TM_LABS_WILDCARD_CERT_ARN"
+
+$PRODUCTION && STACK_NAME="rre" || STACK_NAME="rre-beta"
```
[the remark this answers is not in the document]
Comment (mathcolo) [reply devops/deploy.sh:31]: Austin this is a Wendy's

**Pull request merged**